burr/core/application.py

```diff
@@ -31,7 +32,14 @@
     create_action,
     default,
 )
-from burr.core.state import State
+from burr.core.state import BasicStatePersistence, State
+from burr.lifecycle import (
```

Comment:
You shouldn't need to import these. We should:
1. expose `does_hook(hook_name, is_async)` from `_does_hook` on `LifecycleAdapterSet`
2. Call `does_hook` below when we check `isinstance`

Two reasons:
1. dependencies are circular (already) -- we use `if TYPE_CHECKING` in `lifecycle.internal`. This is necessary, but best to minimize
2. `isinstance` is iffy -- it moves the burden of deciding what an object does from the object to the caller.

Reply:
On second thought you'll want to be a little more flexible -- see the note below.